raft/src/main/java/org/apache/kafka/raft/internals/BatchAccumulator.java

```diff
@@ -0,0 +1,296 @@
+            data,
+            memoryPool,
+            currentBatch.initialBuffer()
+        ));
+        currentBatch = null;
+        startNewBatch();
+    }
+
+    private void startNewBatch() {
+        ByteBuffer buffer = memoryPool.tryAllocate(maxBatchSize);
+        if (buffer != null) {
+            currentBatch = new BatchBuilder<>(
+                buffer,
+                serde,
+                compressionType,
+                nextOffset,
+                time.milliseconds(),
+                false,
+                RecordBatch.NO_PARTITION_LEADER_EPOCH,
+                maxBatchSize
+            );
+        }
+    }
+
+    /**
+     * Check whether there are any batches which need flushing now.
+     *
+     * @param currentTimeMs current time in milliseconds
+     * @return true if there are batches ready to flush, false otherwise
+     */
+    public boolean needsFlush(long currentTimeMs) {
+        return timeUntilFlush(currentTimeMs) <= 0;
+    }
+
+    /**
+     * Check the time remaining until the next needed flush. If the accumulator
+     * is empty, then {@link Long#MAX_VALUE} will be returned.
+     *
+     * @param currentTimeMs current time in milliseconds
+     * @return the delay in milliseconds before the next expected flush
+     */
+    public long timeUntilFlush(long currentTimeMs) {
+        lock.lock();
+        try {
+            lingerTimer.update(currentTimeMs);
+            if (isEmpty()) {
+                return Long.MAX_VALUE;
+            } else {
+                return lingerTimer.remainingMs();
+            }
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    private boolean isEmpty() {
+        lock.lock();
+        try {
+            if (currentBatch != null && currentBatch.nonEmpty()) {
+                return false;
+            } else {
+                return completed.isEmpty();
+            }
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    /**
+     * Get the leader epoch, which is constant for each instance.
+     *
+     * @return the leader epoch
+     */
+    public int epoch() {
+        return epoch;
+    }
+
+    /**
+     * Flush completed batches. The caller is expected to first check whether
+     * a flush is expected using {@link #needsFlush(long)} in order to avoid
+     * unnecessary flushing.
+     *
+     * @return the list of completed batches
+     */
+    public List<CompletedBatch<T>> flush() {
+        lock.lock();
+        try {
+            if (currentBatch != null && currentBatch.nonEmpty()) {
+                completeCurrentBatch();
+            }
+
+            List<CompletedBatch<T>> res = completed;
+            this.completed = new ArrayList<>();
+            return res;
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    /**
+     * Get the number of batches including the one that is currently being
+     * written to (if it exists).
+     *
+     * @return
+     */
+    public int count() {
```

Comment:
Is this going to be used for non-testing code in the future? If it is only going to be for metrics purposes maybe we can allow it to be non thread-safe just to not blocking on other critical paths.

Reply:
I am not sure. We don't have a metric yet, so I thought we might as well start with a thread-safe implementation. If we add a metric in the future, we can probably use an `AtomicInteger` or something and eliminate the locking.